core/services/ocrcommon/data_source.go

```diff
@@ -176,19 +194,94 @@ func (ds *inMemoryDataSource) parse(finalResult pipeline.FinalResult) (*big.Int,
 
 // Observe without saving to DB
 func (ds *inMemoryDataSource) Observe(ctx context.Context, timestamp ocr2types.ReportTimestamp) (*big.Int, error) {
-	_, finalResult, err := ds.executeRun(ctx, ObservationTimestamp{
+	_, trrs, err := ds.executeRun(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	finalResult := trrs.FinalResult(ds.lggr)
+	setEATelemetry(ds, trrs, finalResult, ObservationTimestamp{
 		Round:        timestamp.Round,
 		Epoch:        timestamp.Epoch,
 		ConfigDigest: timestamp.ConfigDigest.Hex(),
 	})
-	if err != nil {
-		return nil, err
-	}
+
 	return ds.parse(finalResult)
 }
 
+// inMemoryDataSourceCache is a time based cache wrapper for inMemoryDataSource.
+// If cache update is overdue Observe defaults to standard inMemoryDataSource behaviour.
+type inMemoryDataSourceCache struct {
+	cacheExpiration time.Duration
+	mu              *sync.RWMutex
+	lastUpdate      atomic.Pointer[time.Time]
+	latestTrrs      pipeline.TaskRunResults
+	latestResult    pipeline.FinalResult
+	*inMemoryDataSource
+}
+
+// isUpdateOverdue atomically checks if update is overdue.
+func (ds *inMemoryDataSourceCache) isUpdateOverdue() bool {
```

Comment:
this seems like unneeded in direction. the updater is on a ticker set by the cache TTL

won't the answer to this isUpdateOverdue always be yes?

Reply:
i'd try to refactor to ~
```
type cache {
	mu sync.Mutex
	vals your_data // lastest data
	err error // lastest error if any

	ds datasource
}

func( c *cache) get() ( vals, error) {
	
	c.mu.rlock()
	err :=  c.err
	c.mu.runlock()
	
	if err != nil {
	// cache is stale, try manual update
	c.update()
}
	c.mu.rlock()
defer c.mu.runlock()
	return c.vals, c.err
}
// synchoronous update to the cache value
func (c *cache) update() {
	c.mu.lock()
	defer c.mu.unlock()
	// use datasource to compute a new value
}

// runs update in polling loop
func (c *cache) poll(ctx) {
	ticker = blah
	
	switch {
	ticker: c.update()
	ctx.Done: return
	}
	}